core/Files.cc

```diff
@@ -113,7 +113,9 @@ unique_ptr<File> File::deepCopy(GlobalState &gs) const {
 }
 
 void File::setFileHash(unique_ptr<const FileHash> hash) {
-    hash_ = move(hash);
+    if (hash_ == nullptr) {
+        hash_ = move(hash);
+    }
```

Comment:
ooc what's the reasoning here? is the update only idempotent because of this check?

Reply:
Idempotent in the sense that, once set, the actual pointer to the hash does not change. That was the intent behind this change. I didn't introduce this to fix a bug, but I saw it in Dmitry's original branch with hash serialization and decided that it was a good change to prevent needlessly invalidating references to the hash.

Other than that, `hash_ == hash` since the `source` of a File can never change.

In practice, the code I've written avoids re-computing hashes for already hashed files.

Reply:
I added a comment.